### PR TITLE
Compare tooltip right offset to viewport right offset (and not width)

### PR DESCRIPTION
**Pull request by Intrepidd**
Hi !

There seems to be a bug on viewport bound check.

Instead of checking that the right offset of the tooltip is not bigger than the right offset of the viewport, you check it's not bigger to the viewport size. This causes wrong offsets being applied in some conditions.

I'm willing to add a unit test if you acknowledge this issue but this will be complicated as qunit tests run in a div with a negative positionning, that will prevent the bug from happening.

Thanks

### CODE REVIEW

**Comment by cvrebert** [comment]
> qunit tests run in a div with a negative positioning, that will prevent the bug from happening.

I think we'd be okay with the new test using an on-screen element. We already have to do something similar for iOS (see 20fb55a840f1b8ae7c3ccaa86acb10a2554b903c).

**Comment by Intrepidd** [comment]
Thanks !

**Comment by cvrebert** [comment]
@twbs-savage Retry

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: a0ef4f9bb5c56e38130dcd21ba626a9c1d29f264
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/55854960

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by Intrepidd** [comment]
I tried to add a test case but did not manage to reproduce it within the js unit tests.

I think the minimality of those test prevent the bug from happening (no tooltip css for instance)

However, here's a jsfiddle demonstrating the issue : https://jsfiddle.net/gc8qkgpg/

Thanks

**Comment by cvrebert** [comment]
> (no tooltip css for instance)

Yeah, generally the approach is to include a minimal amount of CSS (either simple stub CSS or actual tooltip CSS copied over from `bootstrap.css`) in the testcase itself via `<style>` tags in those sorts of cases. For examples, see
* https://github.com/twbs/bootstrap/blob/93bf800e1f00861525688007ae7aaf6f3daeda47/js/tests/unit/tooltip.js#L697
* https://github.com/twbs/bootstrap/blob/93bf800e1f00861525688007ae7aaf6f3daeda47/js/tests/unit/tooltip.js#L958

**Comment by cvrebert** [comment]
@Intrepidd So, can you try adding a unit test again now that I've explained the approach?

**Comment by Intrepidd** [comment]
Sorry for taking so long, and thank you for reminding me about this.

I finally managed to write a spec that fails on master and passes on my branch.

The key is that it can only fail when this style (from bootstrap css) is present :

```css
* {
  box-sizing: border-box;
}
```

In the tests this style was added though :

```css
.tooltip, .tooltip *, .tooltip *:before, .tooltip *:after { box-sizing: border-box; }
```

I replaced the latter by the former in the specs that used it to be more consistent.

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 7f187f90c16ae4aad573d166333bb2d16f63c97f
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/59513555

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by Intrepidd** [comment]
I pushed -f, it broke the build :/

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 6ce6fec465cd3f76fac158a981b40c95d52c91d2
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/59513120

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by cvrebert** [comment]
No worries, a build of your new revision is already in progress.

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 4e18806b07a311aac7c5b2fa5a0eb08dc960b462
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/59513874

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

File: js/tests/unit/tooltip.js [hunk 1]
```diff
@@ -733,6 +733,33 @@ $(function () {
...
+    var $container = $('<div class="container-viewport"/>').appendTo(document.body)
+    var $target = $('<a href="#" rel="tooltip" title="tip">foobar</a>')
+      .appendTo($container)
+      .bootstrapTooltip({
+        viewport: '.container-viewport'
+      })
+
+    $target.bootstrapTooltip('show')
+    var $tooltip = $container.find('.tooltip')
+    assert.strictEqual(Math.round($tooltip.offset().left), $target.position().left + $target.width() / 2 - $tooltip[0].offsetWidth / 2)
```
Comment (cvrebert) [inline js/tests/unit/tooltip.js]: Based on https://travis-ci.org/twbs/bootstrap/jobs/59513871, I think you'll need to round the second value too.

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 13cd2cbd58bb3d0724d8a2bd99fcd0b0793d38d2
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/59516395

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by Intrepidd** [comment]
Yay !

**Comment by Intrepidd** [comment]
Is there anything more I can to to wrap up this PR ?

**Comment by cvrebert** [comment]
Sorry for the delay; the `* { box-sizing: border-box; }` change gave me pause when considering merging this previously. Turns out a less drastic reset (cdb37dce3a85c07790e73b7e2c809860db6b0cad) works fine, but I had to test that out (and also squash everything into a single commit).

Thanks for your contribution!

**Comment by Intrepidd** [comment]
Ok ! Thanks for your time !
